### PR TITLE
Enhancement/overlay-sr

### PR DESCRIPTION
**Pull request by FObermaier**
Overlay with snap-rounding

* https://lin-ear-th-inking.blogspot.com/2020/05/jts-overlay-next-generation.html
* https://github.com/dr-jts/jts/tree/overlay-sr

### CODE REVIEW

**Comment by FObermaier** [comment]
**Note: Work in progress**

**Comment by airbreather** [comment]
I'll look into the API later... having the algorithm do its own snap-rounding is going to be a **very** welcome change, so I'm excited to see this.

**Comment by dr-jts** [comment]
A couple of comments on this port:

* The OverlayNG codebase is still getting a few bug fixes and refactorings, so make sure to track the commits on that branch.
* A recent addition is the [SnappingNoder](https://github.com/dr-jts/jts/blob/overlay-sr/modules/core/src/main/java/org/locationtech/jts/noding/snap/SnappingNoder.java), which provides much better robustness for floating precision overlay operations.  This has enabled the [OverlayNGSnapIfNeeded](https://github.com/dr-jts/jts/blob/overlay-sr/modules/core/src/main/java/org/locationtech/jts/operation/overlayng/OverlayNGSnapIfNeeded.java) class, which uses some heuristics to maximise performance and robustness.  This is the recommended class to use for floating-precision overlay.

**Comment by FObermaier** [comment]
@dr-jts, I'm aware of that and I'm keeping an eye on it. But ATM I'm lacking time.

**Comment by juliusfriedman** [comment]
@dr-jts, Glad to see these improvements.

1 thing might I suggest is would be the ability to limit Operations, e.g. If I only need 1 intersection then I should be able to improve the speed of the IntersectionOp by giving a max number of intersections to find, it would also then be trivial to resume from that index if more are needed.

**Comment by FObermaier** [comment]
@juliusfriedman, why would you want only one intersection? If you need a predicate check possibly one against many I strongly urge you to use the functionality provided in the `NetTopologySuite.Geometries.Prepared` namespace.

**Comment by juliusfriedman** [comment]
> @juliusfriedman, why would you want only one intersection? If you need a predicate check possibly one against many I strongly urge you to use the functionality provided in the `NetTopologySuite.Geometries.Prepared` namespace.

Thank you,

What also about if I want to buffer the 1st clockwise intersection a different way than I want to buffer the first counter clockwise intersection?

I can imagine a stepped algorithm which can step the intersections required and pass them to the buffer op or any other op in the chain where required for the processing required by the application...

Is that something which Prepared can also handle?

**Comment by FObermaier** [comment]
> Is that something which Prepared can also handle?

No, that namespace is for predicates only.

**Comment by FObermaier** [comment]
This is now in sync with the original.
What we need to settle:
* [ ] JTS' `OverlayLabeller` uses `Deque<OverlayEdge>`. There is no interface or class for a deque in .NetStandard. Currently a `LinkedList` is used as replacement, we need to consider pulling in our own IDeque<T> interface and inplementation.
* [ ] JTS' has quite some breaking changes which I avoided by attributing with `Obsolete` and remove internal usage. An example is the move from `Position` and `Quadrant` to geometries namespace.

What remains to be done:
* [ ] Improve xml-test reder to be able to parse and perform tests on newer functionality. I have a strong feeling that this is not working ATM.

**Comment by FObermaier** [comment]
@airbreather, this PR is in the same state as the code merged to master in JTS.
The failing tests are due to missing functionality in our `NetTopologySuite.TestRunner` project.

**Comment by airbreather** [comment]
> @airbreather, this PR is in the same state as the code merged to master in JTS.
> The failing tests are due to missing functionality in our `NetTopologySuite.TestRunner` project and old Xml-test-files.

OK thanks for the heads-up.  I'll do a review of the API, but I won't be able to do a full review of the code... there's a lot there.

Sorry I didn't mention this, but I did want to hold off until it got stable enough to get merged into JTS, since it was still in flux when I made my previous comment on the thread.

**Comment by FObermaier** [comment]
Thanks, develop would be ready to merge, too.

I'd like to change `NetTopologySuite.Operation.OverlayNg` namespace to `NetTopologySuite.Operation.OverlayNG` prior to merging.

**Comment by airbreather** [comment]
re: `Deque<OverlayEdge>`, JTS only calls `isEmpty()`, `addFirst()`, and `removeFirst()`, so we can actually use a `Stack<OverlayEdge>`.

**Comment by dr-jts** [comment]
> re: `Deque<OverlayEdge>`, JTS only calls `isEmpty()`, `addFirst()`, and `removeFirst()`, so we can actually use a `Stack<OverlayEdge>`.

Yes, the `Deque` is being used as a stack (as is apparently Java best practice now).

File: src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs:296]: validate `a` non-null before using it

(we could rely on `RobustScale` to validate for us if we do an earlier comment, mainly just mentioning this here for consistency)

File: src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs:314]: validate `a` non-null before using it

(we could rely on `InherentScale` to validate for us if we do an earlier comment, mainly just mentioning this here for consistency)

File: src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs:330]: validate `a` non-null before using it

(we could rely on `InherentScale` to validate for us if we do an earlier comment, mainly just mentioning this here for consistency)

File: src/NetTopologySuite/Operation/OverlayNG/RingClipper.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/RingClipper.cs:30]: this isn't designed to be inherited from, so let's mark it `sealed`

(We can always unseal it later if that becomes a problem)

File: src/NetTopologySuite/Operation/OverlayNG/RingClipper.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/RingClipper.cs:50]: validate `clipEnv` non-null before using it

File: src/NetTopologySuite/Operation/OverlayNG/RingClipper.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/RingClipper.cs:64]: validate `pts` non-null before using it

File: src/NetTopologySuite/Operation/OverlayNG/RobustClipEnvelopeComputer.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/RobustClipEnvelopeComputer.cs:14]: this isn't designed to be inherited from, so let's mark it `sealed`

(We can always unseal it later if that becomes a problem)

File: src/NetTopologySuite/Operation/OverlayNG/UnaryUnionNG.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/UnaryUnionNG.cs:13]: Instead of a private constructor, let's make this `static`.

File: src/NetTopologySuite/Operation/OverlayNG/UnaryUnionNG.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/OverlayNG/UnaryUnionNG.cs:24]: validate `geom` non-null before using it

File: src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs:64]: instead of this conditional rethrow (which should be `throw;` instead of `throw ex;`), use a filtered catch:

```csharp
catch (TopologyException) when (g0.Dimension == Dimension.Surface && g1.Dimension == Dimension.Surface)
{
    return UnionPolygonsByBuffer(g0, g1);
}
```

File: src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs:50]: `@link` should be `<see cref=`

File: src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs:127]: convert javadoc to xmldoc

File: src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs:107]: can we prefix field names with `_`

File: src/NetTopologySuite/Operation/Union/OverlapUnion.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/Union/OverlapUnion.cs:81]: validate `g0` non-null before using it

File: src/NetTopologySuite/Operation/Union/UnionStrategy.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/Union/UnionStrategy.cs:11]: this isn't designed to be inherited from, so let's mark it `sealed`

(We can always unseal it later if that becomes a problem)

File: src/NetTopologySuite/Operation/Union/UnionStrategy.cs
Comment (airbreather) [inline src/NetTopologySuite/Operation/Union/UnionStrategy.cs:29]: validate `g0` non-null before using it

also `g1`???

or just make this `internal`???

File: src/NetTopologySuite/Planargraph/DirectedEdge.cs
Comment (airbreather) [inline src/NetTopologySuite/Planargraph/DirectedEdge.cs:95]: Not sure about this name... maybe [`QuadrantEx`](https://stackoverflow.com/q/4039054/1083771)?

**Review by airbreather — CHANGES_REQUESTED** [review]
I only looked at the non-test files.

Let me know if you want me to fix all of these items.  I don't mind.

File: src/NetTopologySuite/Geometries/Position.cs
Comment (FObermaier) [reply src/NetTopologySuite/Geometries/Position.cs:7]: What I like about enums is that I don't just see the value but also its meaning.   
Maybe we can mix both, rename `Position` enum to `Position[Index|Value]` and then use that inside Position enum-class.
What is the benefit of an enum-class over a struct in this case?

File: src/NetTopologySuite/Geometries/Quadrant.cs
Comment (FObermaier) [reply src/NetTopologySuite/Geometries/Quadrant.cs:18]: dito Position.

File: src/NetTopologySuite/GeometriesGraph/Depth.cs
Comment (FObermaier) [reply src/NetTopologySuite/GeometriesGraph/Depth.cs:129]: It is not used throughout the code-base. Neither in JTS nor in NTS.

File: src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs
Comment (FObermaier) [reply src/NetTopologySuite/Operation/Union/CascadedPolygonUnion.cs:107]: I did not touch that file, did I?

File: src/NetTopologySuite/Planargraph/DirectedEdge.cs
Comment (FObermaier) [reply src/NetTopologySuite/Planargraph/DirectedEdge.cs:95]: Should be in sync with how we name it in `EdgeEnd`

**Comment by FObermaier** [comment]
> Let me know if you want me to fix all of these items. I don't mind.

That'd be great

File: src/NetTopologySuite/Geometries/Position.cs
Comment (airbreather) [reply src/NetTopologySuite/Geometries/Position.cs:7]: >  What is the benefit of an enum-class over a struct in this case?

It could be a struct instead, the point I wanted to make is just that a .NET `enum` can't have methods or properties, so in trying to port one Java class this way, we wind up making two .NET types, with one of them feeling out of place.

In most places, I also favor designing around .NET `enum`s, but obviously our design is more constrained in core NTS.

File: src/NetTopologySuite/GeometriesGraph/Depth.cs
Comment (airbreather) [reply src/NetTopologySuite/GeometriesGraph/Depth.cs]: ok, just updated the message then

File: src/NetTopologySuite/GeometriesGraph/QuadrantOp.cs
Comment (airbreather) [reply src/NetTopologySuite/GeometriesGraph/QuadrantOp.cs]: removed the private constructor altogether and just made the class `static`

File: src/NetTopologySuite/GeometriesGraph/QuadrantOp.cs
Comment (FObermaier) [inline src/NetTopologySuite/GeometriesGraph/QuadrantOp.cs:15]: This is not a breaking change?

Comment (airbreather) [reply src/NetTopologySuite/GeometriesGraph/QuadrantOp.cs:15]: It shouldn't be -- since `QuadrantOp` can't be instantiated, the only thing you can do with it is call static methods.

I guess it's theoretically possible that someone's doing something like this:
```csharp
QuadrantOp op = null;
```

but that's really far-fetched.

Comment (airbreather) [reply src/NetTopologySuite/GeometriesGraph/QuadrantOp.cs:15]: ahh, the API Compat check actually rejects it, so I'm going to revert.

File: src/NetTopologySuite/Operation/OverlayNG/Edge.cs
Comment (airbreather) [reply src/NetTopologySuite/Operation/OverlayNG/Edge.cs:72]: on second thought, no

File: src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs
Comment (airbreather) [reply src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs:267]: This approach is respectably fast, but there are values that cause it to give incorrect answers (probably dotnet/runtime#1643).

The following returns the same value as what's there, in all the cases I've tested, plus those other cases that failed the above, and it's even faster (for real-world values) than the broken `Math.Round` alternative:
```csharp
public static int NumberOfDecimals(double value)
{
    if (value != Math.Round(value, 15) || value > 79228162514264337593543950335d || value < -79228162514264337593543950335d)
    {
        return NumberOfDecimalsRare(value);
    }

    decimal valueAsDecimal = (decimal)value;
    return ((byte*)&valueAsDecimal)[2];
}
```

File: src/NetTopologySuite/Operation/Union/UnionStrategy.cs
Comment (airbreather) [reply src/NetTopologySuite/Operation/Union/UnionStrategy.cs:29]: just making it `internal`.  it still builds.

**Comment by airbreather** [comment]
> * [ ] Improve xml-test reader to be able to parse and perform tests on newer functionality. I have a strong feeling that this is not working ATM.
>
> The failing tests are due to missing functionality in our `NetTopologySuite.TestRunner` project and old Xml-test-files.

The only issue that I directly noticed here was that we were checking for the old `com.vividsolutions` namespaces in a switch statement where the `default` branch had a `Console.ReadKey` call, so those tests would never terminate.  After fixing that, everything passes (though, as I mentioned, I haven't looked at the code for any of the tests).

There's one more failing test: `Fixing_invalid_polygon_with_Buffer_0_returns_empty_polygon`.  I marked it as `[Category("FailureCase")]` for now, but **just** to make sure that everything else gets a green checkmark from Travis-CI (it does :grinning:).

I left the "enum class" feedback open on `Position` and `Quadrant`, since I'm not sure what you want to do about it.  I personally favor that kind of approach since it seems to give the best of both worlds: there's just one type like JTS has (`Position` / `Quadrant`), and yet we still get to use strong typing so that the methods don't just accept any old `int` like they do in JTS.
- If you're OK with me making the change, let me know and I will.

**Comment by airbreather** [comment]
image

(sorry, mis-clicked)

File: src/NetTopologySuite/Geometries/Position.cs
Comment (FObermaier) [reply src/NetTopologySuite/Geometries/Position.cs]: I refactored `Position` to a struct.

File: src/NetTopologySuite/Geometries/Quadrant.cs
Comment (FObermaier) [reply src/NetTopologySuite/Geometries/Quadrant.cs]: I refactored `Quadrant` to a struct.

File: test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs
Comment (airbreather) [inline test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs:239]: We should still figure out why this test fails... is it a bug upstream?

**Review by airbreather — APPROVED** [review]
I'm fine if you want to merge this huge branch before we figure out that test failure, but we should probably figure it out before the next release.

File: test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs
Comment (dr-jts) [reply test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs:239]: This is because the polygon is actually a figure-8, and buffer(0) often doesn't handle those correctly.

This test probably isn't relevant any more, because `GeometryPrecisionReducer` now uses Snap-Rounding, which does not use buffer(0) to creating valid topology.

Comment (airbreather) [reply test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs:239]: @dr-jts the exception is actually thrown in `GeometryPrecisionReducer` itself, specifically [here](https://github.com/NetTopologySuite/NetTopologySuite/blob/bead06f69f2cfab88dde7df96d2b4b351abeeaf4/src/NetTopologySuite/Operation/OverlayNG/PolygonBuilder.cs#L190).  I haven't tried this in JTS to determine whether this is a problem with the algorithm or with our port.

⌛ if I get the time to get around to it, I'll try it out and report it in JTS if it fails.

Comment (dr-jts) [reply test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs:239]: Ah, I understand now.  JTS throws the same error.  And in fact it throws for any precision value.  The problem is that the input geometry is invalid.  SInce `GeometryPrecisionReducer` now uses OverlayNG, it is more sensitive to invalidity.  This might be considered a regression, I suppose - but it still seems like a reasonable & consistent contract that input needs to be valid.  

The usual heuristic to fix bad geometry is to use `buffer(0)`, but in this case as noted that fails, producing a `POLYGON EMPTY`.  It is only a heuristic, and never guaranteed to work.  At some point hopefully JTS will acquire a `MakeValid` function.

Comment (dr-jts) [reply test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs:239]: The cause of the invalidity is a "bowtie" self-intersection at the right end of the polygon.  The points forming it are only on the order of 1e-15 apart!

image

It is interesting that this fails, since you would expect these points to snap together.  This might actual reveal a flaw in the OverlayNG algorithm, in that it uses the original polygon to check for containment, which leaves it susceptible to input invalidity.  I should probably revisit this design.

**Comment by airbreather** [comment]
:tada:

**Comment by DGuidi** [comment]
👏👏👏

File: test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs
Comment (dr-jts) [reply test/NetTopologySuite.Samples.Console/Tests/Github/Issues.cs:239]: The reason this case fails in `GeometryPrecisionReducer` is because of the invalid input.  Because of the bowtie polygon, the orientation of the ring is ambiguous.  It is evaluated in such a way that the large portion of the ring has orientation indicating that it is a hole.  The smaller "shell" part collapses to a single point, leaving just a hole ring.  This then of course cannot be assigned to a shell.

The throwing of a `TopologyException` is actually a good thing, since it indicates there is something wrong with the input.   It's a bit more worrying that this polygon can actually produce output when used in say intersections (although the output is invalid).  At the moment I don't have an idea of how to improve this.

**Pull request merged**
